polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -362,109 +362,166 @@ struct LoopInternalization
-
-    // Analyze affine memory accesses in the innermost loop.
-    MemorySelector memorySelector(memAccessAnalysis, solver);
-    memorySelector.analyze(*innermostLoop);
-
-    // Determine which memory space should be used for each affine operation.
-    DenseMap<Value, MemorySelector::MemorySpace> valueToMemorySpace;
-    innermostLoop->walk<WalkOrder::PreOrder>([&](Operation *op) {
-      if (!isa<affine::AffineLoadOp, affine::AffineStoreOp>(op))
-        return;
+  std::optional<LoopLikeOpInterface> innermostLoop =
+      LoopTools::getInnermostLoop(loop);
+  assert(innermostLoop.has_value() && "Failed to get the innermost loop");
 
-      affine::MemRefAccess access(op);
-      auto it = valueToMemorySpace.find(access.memref);
-      if (it != valueToMemorySpace.end())
-        return;
+  // Use the memory selector to determine the ideal memory space for memref
+  // accesses in the innermost loop.
+  MemorySelector memorySelector(memAccessAnalysis, solver);
+  memorySelector.analyze(*innermostLoop);
 
-      std::optional<MemorySelector::MemorySpace> memSpace =
-          memorySelector.selectMemorySpace(access.memref);
-      if (memSpace)
-        valueToMemorySpace[access.memref] = *memSpace;
-    });
+  loop->walk<WalkOrder::PreOrder>([&](Operation *op) {
+    if (!isa<affine::AffineLoadOp, affine::AffineStoreOp>(op))
+      return;
+
+    affine::MemRefAccess memRefAccess(op);
+
+    // Skip if the memref is already in the map.
+    if (memrefToMemorySpace.find(memRefAccess.memref) !=
+        memrefToMemorySpace.end())
+      return;
 
-    // TODO: prioritize the array accesses that should use shared memory.
+    // Compute the ideal memory space if possible.
+    std::optional<MemorySelector::MemorySpace> memSpace =
+        memorySelector.selectMemorySpace(memRefAccess.memref);
+    if (!memSpace)
+      return;
 
-    SmallVector<T> nestedLoops;
-    LoopTools::getPerfectlyNestedLoops(nestedLoops, loop);
+    // Record we have processes the memref in this loop...
+    auto it = loopToMemref.find(loop);
+    if (it == loopToMemref.end())
+      loopToMemref[loop] = {memRefAccess.memref};
+    else {
+      SmallVector<Value> &memRefs = loopToMemref[loop];
+      memRefs.push_back(memRefAccess.memref);
+    }
 
-    SmallVector<Value> tileSizes;
-    getTileSizes(nestedLoops, tileSizes);
+    // ... and record the memref memory space.
+    memrefToMemorySpace[memRefAccess.memref] = *memSpace;
+  });
+}
 
-    SmallVector<T> tiledNest;
-    LogicalResult res = tile(nestedLoops, tileSizes, tiledNest);
-    LLVM_DEBUG({
-      if (res.succeeded())
-        llvm::dbgs() << "Tiled loop:\n" << tiledNest.front() << "\n";
-      else
-        llvm::dbgs() << "Tile NOT performed\n";
-    });
+template <typename T>
+void LoopInternalization::transform(
+    T loop, const MemoryAccessAnalysis &memAccessAnalysis,
+    DataFlowSolver &solver) const {
+  static_assert(llvm::is_one_of<T, affine::AffineForOp, scf::ForOp>::value);
```

Comment:
We are no longer using `enable_if_t`, because that doesn't allow having definition out of the class?